### PR TITLE
utils_test.qemu: online hotpluged memory before check memory

### PR DESCRIPTION
**Pull request by xutian**
Hotpluged memory is offline state by default, this commit online
unusable memory(offline state memory) via kernel sys filesystem
if the memory is not online by systemd.

Signed-off-by: Xu Tian <xutian@redhat.com>

### CODE REVIEW

**Comment by xutian** [comment]
@yashkmankad add a step to online hotplug memory systemd not support online hotpluged memory automatically. 
Would you like to have a try.  I tested it on rhel7.2 guest, it works fine.

(env)[root@dhcp-xutian cfg]# cat /root/avocado/job-results/latest/results.json 
{"tests": [{"status": "PASS", "url": "smp_2.4096m.repeat1.run_test.Host_RHEL.7.2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.hotplug_memory.before.vm_system_reset.hotplug.backend_file.policy_default.two", "logfile": "/root/avocado/job-results/job-2016-04-08T15.32-15e7979/test-results/smp_2.4096m.repeat1.run_test.Host_RHEL.7.2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.hotplug_memory.before.vm_system_reset.hotplug.backend_file.policy_default.two/debug.log", "whiteboard": "{'kvm_userspace_version': 'Unknown'}", "end": 1460100855.864089, "logdir": "/root/avocado/job-results/job-2016-04-08T15.32-15e7979/test-results/smp_2.4096m.repeat1.run_test.Host_RHEL.7.2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.hotplug_memory.before.vm_system_reset.hotplug.backend_file.policy_default.two", "start": 1460100753.388205, "test": "smp_2.4096m.repeat1.run_test.Host_RHEL.7.2.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.hotplug_memory.before.vm_system_reset.hotplug.backend_file.policy_default.two", "fail_reason": "None", "time": 102.47588396072388}], "errors": 0, "job_id": "15e797943264a1261c3e32924e26e3c7398e0333", "skip": 0, "time": 102.47588396072388, "debuglog": "/root/avocado/job-results/job-2016-04-08T15.32-15e7979/job.log", "pass": 1, "failures": 0, "total": 1}
(env)[root@dhcp-xutian cfg]#

**Comment by yashkmankad** [comment]
Thanks @xutian 
Will check right away.

**Comment by yashkmankad** [comment]
I applied the commit but it is still failing with the same error for the JeOS image.
Cleber is working on guest OS support for adding JeOS 22 and 23, maybe that will solve the issue.
#439

**Comment by suqinhuang** [comment]
@xutian could you pls provide the ID

**Comment by xutian** [comment]
No internal bug to track this defeat, because internal OS use latest systemd and has fixed this issue. This is a workaround for Old OS.

**Comment by pezhang** [comment]
All cases get the PASS result.

 (1/5) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u2.spice.qcow2.virtio_scsi.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.hotplug_memory.before.vm_system_reset.hotplug.backend_ram.policy_default.two: PASS (92.19 s)
 (2/5) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u2.spice.qcow2.virtio_scsi.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.hotplug_memory.before.vm_system_reset.hotplug.backend_ram.policy_bind.two: PASS (95.82 s)
 (3/5) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u2.spice.qcow2.virtio_scsi.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.hotplug_memory.before.vm_system_reset.hotplug.backend_file.policy_default.two: PASS (102.27 s)
 (4/5) smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u2.spice.qcow2.virtio_scsi.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.hotplug_memory.before.vm_system_reset.hotplug.backend_file.policy_bind.two: PASS (103.40 s)
 (5/5) smp_4.4096m.repeat1.post_check.Host_RHEL.m7.u2.spice.qcow2.virtio_scsi.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.check_coredump: PASS (70.56 s)
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /root/avocado/job-results/job-2016-05-13T08.45-1651b49/html/results.html
TIME       : 464.23 s

File: virttest/utils_test/qemu.py [hunk 1]
```diff
@@ -1774,9 +1774,49 @@ def check_memory(self, vm=None):
         self.os_type = self.params.get("os_type")
         guest_mem_size = super(MemoryHotplugTest, self).get_guest_total_mem(vm)
         vm_mem_size = self.get_vm_mem(vm)
-        threshold = vm_mem_size * 0.06
+        default_threshold = vm_mem_size * 0.06
+        threshold = float(self.params.get("threshold", default_threshold))
         if abs(guest_mem_size - vm_mem_size) > threshold:
```
Comment (pezhang) [inline virttest/utils_test/qemu.py]: Boot a guest with 4096MB memory, it's actual memory is 3791MB.  
abs (3791 - 4096) > 4096 * 0.06       ( 305 > 245.76)
For purely guest which means before doing hotplug, this step has failed.

Comment (pezhang) [inline virttest/utils_test/qemu.py]: My test result: smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u2.spice.qcow2.virtio_scsi.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.hotplug_memory_repeat.backend_file.policy_default: FAIL (59.25 s)

The attachment file is it's debug.log.

Comment (xutian) [inline virttest/utils_test/qemu.py]: Please check how much memory used by your video card,  I guest this memory address obligate by video card.

Comment (xutian) [inline virttest/utils_test/qemu.py]: And if I'm right, please configure 'threshold' param to increase or decrease obligate address. thanks.

**Comment by xutian** [comment]
@pezhang PR updated, would you like to review it. thanks

**Comment by suqinhuang** [comment]
@xutian hi, could you update the patch as CI error.

**Comment by xutian** [comment]
@suqinhuang  updated. thanks your reminder.

**Comment by pezhang** [comment]
@xutian  I tested this patch with https://github.com/autotest/tp-qemu/pull/611. 

After run 106 hotplug/hotunplug loop, the case failed because of hot unplug failing. It's issue of hot-unplug(existed: 1245892), not this patch. As this patch works well with 106 hot plug, so I think this patch works well. 

Resutls:
smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u2.qcow2.virtio_scsi.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.hotplug_memory_repeat.backend_ram.policy_default: ERROR (3514.89 s)

# cat debug.log
10470 16:45:33 INFO | Context: Hotplug/unplug loop '107' --> offline memory40 in guest OS
10471 16:45:34 DEBUG| Sending command: echo 'offline' > /sys/devices/system/memory/memory40/state
...
10527 16:46:00 ERROR| ERROR 1-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u2.qcow2.virtio_scsi.up.virtio_net.RHEL.7.2.x8      6_64.io-github-autotest-qemu.hotplug_memory_repeat.backend_ram.policy_default -> ShellCmdError: Shell command fa      iled: "echo 'offline' > /sys/devices/system/memory/memory40/state"    (status: 1,    output: "echo 'offline' > /      sys/devices/system/memory/memory40/s< 'offline' > /sys/devices/system/memory/memory40/st                               \x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08ate\n-bash:       echo: write error: Device or resource busy\n")

debug.log info:
[debug.txt](https://github.com/avocado-framework/avocado-vt/files/334443/debug.txt)

**Comment by xutian** [comment]
@pezhang right. it's really a product bug and thanks you test.

**Pull request merged**